### PR TITLE
DOSE-518 cache-healing read from object store (kernel part only)

### PR DESCRIPTION
**Pull request by prakashsurya**
### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by prakashsurya** [comment]
@grwilson @ahrens is this the sort of change you were thinking? I feel like there's got to be more to it.. but figured I'd run this by you two..

**Comment by grwilson** [comment]
> @grwilson @ahrens is this the sort of change you were thinking? I feel like there's got to be more to it.. but figured I'd run this by you two..

Yes, this is what I was thinking.

**Comment by prakashsurya** [comment]
OK, thanks for the confirmation. I'll set up an environment to verify this.

**Comment by prakashsurya** [comment]
@ahrens @grwilson I tested this in conjunction with #439 and verified that after corrupting the cache device with `dd`, both a `zpool scrub`, and reading a file from the corrupted pool/cache, results in the agent healing those blocks.

I verified this by:

1. observing the "read block" requests, and seeing the "heal" flag set:
```
$ grep "read block" out | tail
[2021-09-15 15:55:31.242][zettaobject::kernel_connection][TRACE] got request: {"Type": Str("read block"), "size": Uint64(131072), "block": Uint64(731), "heal": BoolV(true), "request_id": Uint64(731), "token": Uint64(18446627044280887104)}
[2021-09-15 15:55:31.242][zettaobject::kernel_connection][TRACE] got request: {"Type": Str("read block"), "size": Uint64(131072), "block": Uint64(732), "heal": BoolV(true), "request_id": Uint64(732), "token": Uint64(18446627044280877056)}
[2021-09-15 15:55:31.242][zettaobject::kernel_connection][TRACE] got request: {"Type": Str("read block"), "size": Uint64(131072), "block": Uint64(733), "heal": BoolV(true), "request_id": Uint64(733), "token": Uint64(18446627044280890872)}
[2021-09-15 15:55:31.242][zettaobject::kernel_connection][TRACE] got request: {"Type": Str("read block"), "size": Uint64(131072), "block": Uint64(735), "heal": BoolV(true), "request_id": Uint64(735), "token": Uint64(18446627044280892128)}
[2021-09-15 15:55:31.242][zettaobject::kernel_connection][TRACE] got request: {"Type": Str("read block"), "size": Uint64(131072), "block": Uint64(734), "heal": BoolV(true), "request_id": Uint64(734), "token": Uint64(18446627044427472240)}
[2021-09-15 15:55:31.242][zettaobject::kernel_connection][TRACE] got request: {"Type": Str("read block"), "size": Uint64(131072), "block": Uint64(736), "heal": BoolV(true), "request_id": Uint64(736), "token": Uint64(18446627042568786848)}
[2021-09-15 15:55:31.242][zettaobject::kernel_connection][TRACE] got request: {"Type": Str("read block"), "size": Uint64(131072), "block": Uint64(737), "heal": BoolV(true), "request_id": Uint64(737), "token": Uint64(18446627042568788104)}
[2021-09-15 15:55:31.242][zettaobject::kernel_connection][TRACE] got request: {"Type": Str("read block"), "size": Uint64(131072), "block": Uint64(738), "heal": BoolV(true), "request_id": Uint64(738), "token": Uint64(18446627042568795640)}
[2021-09-15 15:55:31.252][zettaobject::kernel_connection][TRACE] got request: {"Type": Str("read block"), "size": Uint64(131072), "block": Uint64(739), "heal": BoolV(true), "request_id": Uint64(739), "token": Uint64(18446627042568784336)}
[2021-09-15 15:55:31.253][zettaobject::kernel_connection][TRACE] got request: {"Type": Str("read block"), "size": Uint64(131072), "block": Uint64(740), "heal": BoolV(true), "request_id": Uint64(740), "token": Uint64(18446627042568790616)}
```

2. observing "Healing" messages in the log:
```
$ grep "Healing" out | tail
[2021-09-15 15:55:31.355][zettacache::zettacache][DEBUG] Healing by overwriting: IndexKey { guid: PoolGuid(3010980111213399344), block: BlockId(733) } IndexValue { location: DiskLocation { offset: 1757208064 }, size: 131072, atime: Atime(14361) }
[2021-09-15 15:55:31.355][zettacache::zettacache][DEBUG] Healing by overwriting: IndexKey { guid: PoolGuid(3010980111213399344), block: BlockId(736) } IndexValue { location: DiskLocation { offset: 1757470208 }, size: 131072, atime: Atime(14361) }
[2021-09-15 15:55:31.356][zettacache::zettacache][DEBUG] Healing by overwriting: IndexKey { guid: PoolGuid(3010980111213399344), block: BlockId(731) } IndexValue { location: DiskLocation { offset: 1756945920 }, size: 131072, atime: Atime(14361) }
[2021-09-15 15:55:31.356][zettacache::zettacache][DEBUG] Healing by overwriting: IndexKey { guid: PoolGuid(3010980111213399344), block: BlockId(739) } IndexValue { location: DiskLocation { offset: 1756290560 }, size: 131072, atime: Atime(14361) }
[2021-09-15 15:55:31.356][zettacache::zettacache][DEBUG] Healing by overwriting: IndexKey { guid: PoolGuid(3010980111213399344), block: BlockId(719) } IndexValue { location: DiskLocation { offset: 1755110912 }, size: 131072, atime: Atime(14361) }
[2021-09-15 15:55:31.357][zettacache::zettacache][DEBUG] Healing by overwriting: IndexKey { guid: PoolGuid(3010980111213399344), block: BlockId(732) } IndexValue { location: DiskLocation { offset: 1757076992 }, size: 131072, atime: Atime(14361) }
[2021-09-15 15:55:31.363][zettacache::zettacache][DEBUG] Healing by overwriting: IndexKey { guid: PoolGuid(3010980111213399344), block: BlockId(737) } IndexValue { location: DiskLocation { offset: 1757732352 }, size: 131072, atime: Atime(14361) }
[2021-09-15 15:55:31.375][zettacache::zettacache][DEBUG] Healing by overwriting: IndexKey { guid: PoolGuid(3010980111213399344), block: BlockId(730) } IndexValue { location: DiskLocation { offset: 1756683776 }, size: 131072, atime: Atime(14361) }
[2021-09-15 15:55:31.382][zettacache::zettacache][DEBUG] Healing by overwriting: IndexKey { guid: PoolGuid(3010980111213399344), block: BlockId(729) } IndexValue { location: DiskLocation { offset: 1756552704 }, size: 131072, atime: Atime(14361) }
[2021-09-15 15:55:31.394][zettacache::zettacache][DEBUG] Healing by overwriting: IndexKey { guid: PoolGuid(3010980111213399344), block: BlockId(725) } IndexValue { location: DiskLocation { offset: 1755766272 }, size: 131072, atime: Atime(14361) }
```

3. observing the "head_by" metrics in the log:
```
    healed_by_overwriting: ZettaCacheMetricsHealedByOverwriting {
        hit_count: HitCount(
            2074,
        ),
    },
    healed_by_evicting: ZettaCacheMetricsHealedByEvicting {
        hit_count: HitCount(
            0,
        ),
    },
```

4. additionally, I tried a second "zpool scrub", and/or reading the file a second time, and did not see any further "Healing" messages, nor did the "head_by" metrics increase; meaning the cache was properly healed the first time.

**Comment by prakashsurya** [comment]
One thing worth noting, when doing a "zpool scrub", even if data from the cache was healed, this won't be reported by "zpool status"; e.g.
```
  scan: scrub repaired 0B in 00:00:04 with 0 errors on Wed Sep 15 16:01:28 2021
```
so, even though data was healed, it shows `0B`  as repaired; not sure if that's correct or not, but seems a bit misleading.

**Review by ahrens — APPROVED** [review]

**Comment by prakashsurya** [comment]
Additionally, I verified pool import still works, even with a corrupted cache device; e.g. export pool, corrupt cache, import pool, verify healing. And also, verified "zpool status" when reading a corrupted file shows the checksum errors; e.g.:
```
  pool: test
 state: ONLINE
status: One or more devices has experienced an unrecoverable error.  An
        attempt was made to correct the error.  Applications are unaffected.
action: Determine if the device needs to be replaced, and clear the errors
        using 'zpool clear' or replace the device with 'zpool replace'.
   see: https://openzfs.github.io/openzfs-docs/msg/ZFS-8000-9P
  scan: scrub repaired 0B in 00:00:04 with 0 errors on Wed Sep 15 16:09:31 2021
config:

        NAME                 STATE     READ WRITE CKSUM
        test                 ONLINE       0     0     0
          cloudburst-data-2  ONLINE       0     0    17

errors: No known data errors
```

File: module/os/linux/zfs/vdev_object_store.c [hunk 1]
```diff
@@ -408,6 +409,12 @@ agent_io_block_alloc(zio_t *zio)
 	}
 	fnvlist_add_uint64(nv, AGENT_SIZE, zio->io_size);
 	fnvlist_add_uint64(nv, AGENT_BLKID, blockid);
+
+	if ((zio->io_flags & ZIO_FLAG_IO_RETRY) ||
+	    (zio->io_flags & ZIO_FLAG_SCRUB)) {
+		fnvlist_add_boolean_value(nv, AGENT_HEAL, 1);
```
Comment (grwilson) [inline module/os/linux/zfs/vdev_object_store.c:415]: Can we not use B_TRUE?

Comment (prakashsurya) [reply module/os/linux/zfs/vdev_object_store.c]: I believe we can; I can do that.

**Review by grwilson — APPROVED** [review]

**Pull request merged**